### PR TITLE
Passkey (Webauthn) verification

### PR DESCRIPTION
**Pull request by nalinbhardwaj**
https://github.com/daimo-eth/daimo/assets/6984346/900eaef7-1cb1-4ce7-a419-7e2293e9de1a

### CODE REVIEW

File: apps/daimo-mobile/src/logic/passkey.ts [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+  getAbiItem,
+  hexToBytes,
+} from "viem";
+
+import { Log } from "./log";
+
+require("cbor-rn-prereqs"); // Make CBOR work in React Native
+const cbor = require("cbor");
+
+function parseMakeCredAuthData(buffer: Uint8Array) {
```
Comment (dcposch) [inline apps/daimo-mobile/src/logic/passkey.ts:20]: i'd add a one line comment summarizing what this does + link to spec

File: apps/daimo-mobile/src/logic/passkey.ts [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    // Get Challenge assertion
+    const {
+      derSig,
+      accountName,
+      keySlot,
+      rawAuthenticatorData,
+      clientDataJSON,
+      challengeLocation,
+      responseTypeLocation,
+    } = await requestPasskeySignature(challengeB64, "daimo.xyz");
```
Comment (dcposch) [inline apps/daimo-mobile/src/logic/passkey.ts:124]: pull `daimo.xyz` into a constant eg

```
const passkeyDomain = process.env.DAIMO_PASSKEY_DOMAIN || '';
```

File: apps/daimo-mobile/src/view/screen/AddPasskeyScreen.tsx [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      case "loading":
+        return <ActivityIndicator size="large" />;
+      case "success":
+        return <ButtonBig type="success" title="Success" disabled />;
+      case "error":
+        return <ButtonBig type="danger" title="Error" disabled />;
+    }
+  })();
+
+  const cloudName =
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/screen/AddPasskeyScreen.tsx:83]: 🚀

File: apps/daimo-mobile/src/view/screen/AddPasskeyScreen.tsx [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    Platform.OS === "ios" ? "iCloud Keychain" : "Google Password Manager";
+
+  return (
+    <View style={styles.vertOuter}>
+      <TextCenter>
+        <TextH1>Backup account</TextH1>
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/screen/AddPasskeyScreen.tsx:89]: maybe

> **Passkey backup**
>
> Save a secure passkey in {cloudName}. This way, your funds will be safe even if you lose your device.

File: apps/daimo-mobile/src/view/screen/OnboardingScreen.tsx [hunk 5]
```diff
@@ -458,25 +484,158 @@ function UseExistingPage({
...
+    assert(pubKeyHex !== undefined);
+    console.log(`[ONBOARDING] restore backup attempt`);
+    setRestoreStatus({ status: "loading", message: "Requesting backup" });
+
+    const bChallenge = hexToBytes("0xdead");
+    const challengeB64 = base64.encode(bChallenge);
+    try {
+      const { accountName } = await requestPasskeySignature(
+        challengeB64,
+        "daimo.xyz"
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/screen/OnboardingScreen.tsx:526]: ya def worth extracting, occurs in lots of spots

File: apps/daimo-mobile/src/view/screen/OnboardingScreen.tsx [hunk 6]
```diff
@@ -494,6 +653,22 @@ function UseExistingPage({
...
+          <Spacer h={32} />
+          <TextCenter>
+            <TextLight>or</TextLight>
+          </TextCenter>
+          <Spacer h={32} />
+          {restoreStatus.status !== "success" && restoreElement}
+          {restoreStatus.status === "success" &&
+            account !== null &&
+            nextSlot !== null &&
+            addDeviceElement}
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/screen/OnboardingScreen.tsx:671]: 🚀

File: packages/contract/foundry.toml [hunk 7]
```diff
@@ -5,7 +5,6 @@ libs = ['lib']
 
 # Compiler settings
 solc = "0.8.21"
-via-ir = true
```
Comment (dcposch) [inline packages/contract/foundry.toml:8]: 👍

**Review by dcposch — COMMENTED** [review]
Looks fantastic.

We might want to split this into two PRs:
- The contract side, we can land now with a disclaimer.
- The UI side, later post-`daimo.com`

File: apps/daimo-mobile/index.js [hunk 8]
```diff
@@ -4,7 +4,7 @@ import { registerRootComponent } from "expo";
 import "react-native-get-random-values"; // needs to precede ethers
 
 import "@ethersproject/shims";
-import "fast-text-encoding";
+import "text-encoding-polyfill";
```
[the remark this answers is not in the document]
Comment (nalinbhardwaj) [reply apps/daimo-mobile/index.js:7]: it seems ethers also uses this (thats why we had the polyfill there before)

**Comment by nalinbhardwaj** [comment]
Closing for #383 and #343

**Pull request closed without merging**